### PR TITLE
Expose final internals to allow decoupling Drawing.

### PR DESCRIPTION
**Pull request by JimBobSquarePants**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [x] I have verified that I am following matches the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description
Fixes accessibility for any remaining blockers that are stopping us decoupling ImageSharp.Drawing

TODO:

- [x] `'ImageFrame.Configuration' is inaccessible due to its protection level`.
- [x] `'float' does not contain a definition for 'Clamp' and no accessible extension method 'Clamp' accepting a first argument of type 'float'`.
- [x] `'Image<TPixel>' does not contain a definition for 'GetMemoryAllocator' and no accessible extension method 'GetMemoryAllocator' accepting a first argument of type 'Image<TPixel>' could be found`
- [x] `'MemoryAllocator' does not contain a definition for 'Allocate2D' and no accessible extension method 'Allocate2D' accepting a first argument of type 'MemoryAllocator' could be found`.
- [x] `'GraphicsOptions.IsOpaqueColorWithoutBlending(Color)' is inaccessible due to its protection level`

I consider this issue the highest priority for now and would really appreciate any assistance that can be given determining the best fix for each outstanding issue.

### CODE REVIEW

File: src/ImageSharp.Drawing/Common/Extensions/GraphicsOptionsExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        /// </summary>
+        /// <param name="options">The graphics options.</param>
+        /// <param name="color">The source color.</param>
+        /// <returns>true if the color can be considered opaque</returns>
+        /// <remarks>
+        /// Blending and composition is an expensive operation, in some cases, like
+        /// filling with a solid color, the blending can be avoided by a plain color replacement.
+        /// This method can be useful for such processors to select the fast path.
+        /// </remarks>
+        public static bool IsOpaqueColorWithoutBlending(this GraphicsOptions options, Color color)
```
Comment (JimBobSquarePants) [inline src/ImageSharp.Drawing/Common/Extensions/GraphicsOptionsExtensions.cs]: Why are `TextGraphicsOptions` and `GraphicsOptions` structs? They're both mutable and likely cause unnecessary copying. This extension method is actually copying the struct which cannot be avoided just now.

**Comment by JimBobSquarePants** [comment]
Ok. I'm managed workarounds for everything but the `MemoryAllocator` limitations. @antonfirsov I'll need your help there.

**Comment by JimBobSquarePants** [comment]
I've temporarily enabled internals to test the build cross platform. To test and develop locally comment out the following line in the `ImageSharp.csproj`

``` xml
<InternalsVisibleTo Include="SixLabors.ImageSharp.Drawing" />
```

**Comment by JimBobSquarePants** [comment]
@antonfirsov Is there any reason you can think of why we shouldn't expose the first two extension methods here?

https://github.com/SixLabors/ImageSharp/blob/41c76515f7e8aeda9a2e7c4a1efa52fd9a3f5ac6/src/ImageSharp/Memory/MemoryAllocatorExtensions.cs#L14

File: src/ImageSharp.Drawing/Common/Extensions/GraphicsOptionsExtensions.cs (hunk 1 above)
Comment (JimBobSquarePants) [reply src/ImageSharp.Drawing/Common/Extensions/GraphicsOptionsExtensions.cs]: I've converted them to classes.

File: src/ImageSharp.Drawing/Processing/GradientBrush.cs [hunk 2]
```diff
@@ -121,19 +117,8 @@ internal abstract class GradientBrushApplicator<TPixel> : BrushApplicator<TPixel
...
-                        // TODO: this should be changeble for different gradienting functions
-                        Vector4 result = PorterDuffFunctions.NormalSrcOver(
-                            fromAsVector,
-                            toAsVector,
-                            onLocalGradient);
-
-                        TPixel resultColor = default;
-                        resultColor.FromVector4(result);
-                        return resultColor;
+                        return new Color(Vector4.Lerp((Vector4)from.Color, (Vector4)to.Color, onLocalGradient)).ToPixel<TPixel>();
```
Comment (JimBobSquarePants) [inline src/ImageSharp.Drawing/Processing/GradientBrush.cs]: This fixes #1040

File: src/ImageSharp.Drawing/Utils/NumberUtils.cs [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+namespace SixLabors.ImageSharp
+{
+    /// <summary>
+    /// Utility methods for numeric primitives.
+    /// </summary>
+    internal static class NumberUtils
+    {
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static float ClampFloat(float value, float min, float max)
```
[the remark this answers is not in the document]
Comment (JimBobSquarePants) [reply src/ImageSharp.Drawing/Utils/NumberUtils.cs]: We actually covered that ages ago 😄 

https://github.com/SixLabors/ImageSharp/blob/41c76515f7e8aeda9a2e7c4a1efa52fd9a3f5ac6/tests/ImageSharp.Benchmarks/General/BasicMath/ClampFloat.cs#L8

File: src/ImageSharp.Drawing/Processing/RadialGradientBrush.cs [hunk 4]
```diff
@@ -35,16 +35,18 @@ public sealed class RadialGradientBrush : GradientBrush
...
             new RadialGradientBrushApplicator<TPixel>(
+                configuration,
                 source,
-                options,
                 this.center,
                 this.radius,
                 this.ColorStops,
-                this.RepetitionMode);
+                this.RepetitionMode,
+                options);
```
[the remark this answers is not in the document]
Comment (JimBobSquarePants) [reply src/ImageSharp.Drawing/Processing/RadialGradientBrush.cs]: It was all over the place before I refactored. Happy to move it.

**Comment by JimBobSquarePants** [comment]
Thanks for the fast review @antonfirsov I'm keen to move forward with this.

> GraphicsOptions.Default and TextGraphicsOptions.Default are exposing an arbitrary shared mutable state. It was not the case when those were struct-s, and I think we should avoid this, and replace the singletons with default constructor calls.
> 1. The classes need a deep copy logic, just like Configuration
> 2. The classes are showing the symptoms of the Telescoping Constructor anti-pattern. These constructors are unnecessary, since the properties are settable.

Agree with all of the above. That won't take me long.

>Additionally, we should remove <InternalsVisibleTo Include="SixLabors.ImageSharp.Drawing" /> from ImageSharp.csproj to ensure this is indeed the final step to close #1002.

Yeah, good point, will do.

File: src/ImageSharp.Drawing/Utils/NumberUtils.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (JimBobSquarePants) [reply src/ImageSharp.Drawing/Utils/NumberUtils.cs]: I'm glad the current `Clamp()` extensions methods are internal actually. I'd like to make them explicit like this in the future, much better readability.

[the remark this answers is not in the document]
Comment (antonfirsov) [reply src/ImageSharp.Drawing/Utils/NumberUtils.cs]: That benchmark is actually using `Math.Clamp()`, which does a conversion internally (for what reason ?? 😮), which explains it entirely. `MathF` has no `Clamp` method actually.

**Comment by JimBobSquarePants** [comment]
@antonfirsov I have updated everything except the csproj `InternalsVisibleTo` declaration as the shared infrastructor classes used in Drawing rely on that call until we can move the library to a separate repository.

File: src/ImageSharp.Drawing/Processing/TextGraphicsOptions.cs [hunk 5]
```diff
@@ -9,120 +10,151 @@ namespace SixLabors.ImageSharp.Processing
...
-
-            this.antialiasSubpixelDepth = 16;
-            this.ColorBlendingMode = PixelColorBlendingMode.Normal;
-            this.AlphaCompositionMode = PixelAlphaCompositionMode.SrcOver;
-            this.blendPercentage = 1;
-            this.antialias = enableAntialiasing;
-            this.dpiX = DefaultTextDpi;
-            this.dpiY = DefaultTextDpi;
-        }
+        public static TextGraphicsOptions Default { get; } = new TextGraphicsOptions();
```
Comment (antonfirsov) [inline src/ImageSharp.Drawing/Processing/TextGraphicsOptions.cs]: The singleton is still there. It's completely unnecessary IMO. Shared mutable state that is quite arbitrary. We can replace it with `new TextGraphicsOptions()` on all call sites. Same for `GraphicsOptions`.

File: tests/ImageSharp.Tests/Drawing/Text/TextGraphicsOptionsTests.cs [hunk 6]
```diff
@@ -9,25 +11,139 @@ namespace SixLabors.ImageSharp.Tests.Drawing.Text
 {
     public class TextGraphicsOptionsTests
     {
+        private readonly TextGraphicsOptions newTextGraphicsOptions = new TextGraphicsOptions();
+        private readonly TextGraphicsOptions defaultTextGraphicsOptions = TextGraphicsOptions.Default;
+        private readonly TextGraphicsOptions cloneTextGraphicsOptions = TextGraphicsOptions.Default.Clone();
```
Comment (antonfirsov) [inline tests/ImageSharp.Tests/Drawing/Text/TextGraphicsOptionsTests.cs]: We need proper tests for `Clone()`, which are actually altering the state of the instance before cloning it.

Comment (JimBobSquarePants) [reply tests/ImageSharp.Tests/Drawing/Text/TextGraphicsOptionsTests.cs]: Done.

File: src/ImageSharp.Drawing/Processing/TextGraphicsOptions.cs (hunk 5 above)
Comment (JimBobSquarePants) [reply src/ImageSharp.Drawing/Processing/TextGraphicsOptions.cs]: Ah, didn't get your point previously. Removed.

**Comment by JimBobSquarePants** [comment]
> I still think we shall get rid of those build errors in this step, or a separate PR before actually moving the whole thing out. The Drawing project shall include the shared infrastructure independently.

Feel free to try. I got an ambiguity errors when I tried.

**Pull request merged**